src/linear_algebra/basis.lean

```diff
@@ -751,6 +751,51 @@ end
 
 end span
 
+lemma smul_group_span_eq_top
+  {G : Type*} [group G] [distrib_mul_action G R] [distrib_mul_action G M]
+  [is_scalar_tower G R M] {v : ι → M} (hv : submodule.span R (set.range v) = ⊤) {w : ι → G} :
+  submodule.span R (set.range (w • v)) = ⊤ :=
+begin
+  rw eq_top_iff,
+  intros j hj,
+  rw ← hv at hj,
+  rw submodule.mem_span at hj ⊢,
+  refine λ p hp, hj p (λ u hu, _),
+  obtain ⟨i, rfl⟩ := hu,
+  have : ((w i)⁻¹ • 1 : R) • w i • v i ∈ p := p.smul_mem ((w i)⁻¹ • 1 : R) (hp ⟨i, rfl⟩),
+  rwa [smul_one_smul, inv_smul_smul] at this,
+end
+
+/-- Given a basis `v` and a map `w` such that for all `i`, `w i` are elements of a group,
+`smul_group` provides the basis corresponding to `w • v`. -/
+def smul_group {G : Type*} [group G] [distrib_mul_action G R] [distrib_mul_action G M]
+  [is_scalar_tower G R M] [smul_comm_class G R M] (v : basis ι R M) (w : ι → G) :
+  basis ι R M :=
+@basis.mk ι R M (w • v) _ _ _
+  (linear_independent.group_smul v.linear_independent w) (smul_group_span_eq_top v.span_eq)
+
+lemma smul_group_apply {G : Type*} [group G] [distrib_mul_action G R] [distrib_mul_action G M]
+  [is_scalar_tower G R M] [smul_comm_class G R M] {v : basis ι R M} {w : ι → G} (i : ι) :
+  v.smul_group w i = (w • v : ι → M) i :=
+mk_apply
+  (linear_independent.group_smul v.linear_independent w) (smul_group_span_eq_top v.span_eq) i
+
+lemma smul_of_is_unit_span_eq_top {v : ι → M} (hv : submodule.span R (set.range v) = ⊤)
+  {w : ι → units R} : submodule.span R (set.range (w • v)) = ⊤ :=
+smul_group_span_eq_top hv
+
+/-- Given a basis `v` and a map `w` such that for all `i`, `w i` is a unit, `smul_of_is_unit`
+provides the basis corresponding to `w • v`. -/
+def smul_of_is_unit (v : basis ι R M) (w : ι → units R) :
+  basis ι R M :=
+@basis.mk ι R M (w • v) _ _ _
+  (linear_independent.units_smul v.linear_independent w) (smul_of_is_unit_span_eq_top v.span_eq)
```

Comment:
```suggestion
/-- Given a basis `v` and a map `w` such that for all `i`, `w i` is a unit, `smul_of_is_unit`
provides the basis corresponding to `w • v`. -/
def smul_units (v : basis ι R M) (w : ι → units R) :
  basis ι R M :=
@basis.mk ι R M (w • v) _ _ _
  (linear_independent.units_smul v.linear_independent w) (smul_of_is_unit_span_eq_top v.span_eq)
```
I think it might be worth also having a version in terms of `is_unit (w i)` too though.

Reply:
`units_smul` is probably more accurate since we renamed `smul_of_is_unit_linear_independent` to `linear_independent.units_smul`